io/src/tim_grabber.cpp

```diff
@@ -0,0 +1,238 @@
+  float angle = angle_start_;
+
+  cos_dynamic_lookup_table_.reserve (amount_of_data_);
+  cos_dynamic_lookup_table_.clear ();
+  sin_dynamic_lookup_table_.reserve (amount_of_data_);
+  sin_dynamic_lookup_table_.clear ();
+
+  for (std::size_t i = 0; i < amount_of_data_; i++, angle += angle_step)
+  {
+    cos_dynamic_lookup_table_.push_back(std::cos (angle));
+    sin_dynamic_lookup_table_.push_back(std::sin (angle));
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+void
+pcl::TimGrabber::updateLookupTables () {
+  //// note: cos, sin are the same size
+  if (cos_dynamic_lookup_table_.size () != amount_of_data_)
+    buildLookupTables ();
+}
+
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+bool
+pcl::TimGrabber::isValidPacket () const {
+  return received_packet_.data ()[length_-1] == '\03';
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+void
+pcl::TimGrabber::receiveTimPacket ()
+{
+  if (!tim_socket_.is_open ())
+    return;
+
+  tim_socket_.send (boost::asio::buffer ("\x02sRN LMDscandata 1\x03\0"));
+
+  std::this_thread::sleep_for (std::chrono::milliseconds (wait_time_milliseconds_));
+
+  length_ = tim_socket_.receive (boost::asio::buffer (received_packet_));
+
+  if (!isValidPacket ()) {
+    //// Countup wait_time
+    //// Because fit wait_time to network latency
+    wait_time_milliseconds_++;
+
+    //// Clear invalid packet in socket
+    while (!isValidPacket ())
+      length_ = tim_socket_.receive (boost::asio::buffer (received_packet_));
+
+    //// If received packet is invalid, recurse
+    receiveTimPacket ();
+  }
+}
+
+void
+pcl::TimGrabber::parsePacketHeader (std::string const& header) {
+  //// note:  Please following part, if you want to use other information in the header.
+  //std::size_t pos = header.find (' ');
+  //for (int i=0; i<24; ++i)
+    //pos = header.find (' ', pos+1);
+
+  std::size_t pos = header.rfind(' ');
+  amount_of_data_ = std::stoi (header.substr (pos+1), nullptr, 16);
+}
+
+void
+pcl::TimGrabber::parsePacketBody (std::string const& body) {
+  std::istringstream ss (body);
+  ss.setf(std::ios::hex, std::ios::basefield);
+  int d;
+  for (auto& distance : distances_) {
+    ss >> d;
+    distance = d / 1000.0;
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+void
+pcl::TimGrabber::processTimPacket (std::string const& packet)
+{
+  std::size_t header_body_splitter = packet.find (' ');
+  for (int i=0; i<25; ++i)
+    header_body_splitter = packet.find (' ', header_body_splitter + 1);
+
+  //// packet contains header and body.
+  //// header has 26 spaces, the rest is body
+  std::string header (packet.substr (0, header_body_splitter));
+  std::string body   (packet.substr (header_body_splitter + 1));
```

Comment:
Thoughts on using string_view here?

It's in Boost since 1.62 at least (18.04 LTS is more modern than that)

Reply:
Thanks for experimenting.